### PR TITLE
Remove Authorization header regardless of case when redirecting to cross-site

### LINKED ISSUE #1510: authorization header be forwarded to cross-site when redirecting
#1346 fixes only "Authorization" header. "authorization" header isn't supported.

RFC7230 section 3.2 "Header Fields" (https://tools.ietf.org/html/rfc7230#section-3.2) says the following.

> Each header field consists of a case-insensitive field name followed
   by a colon (":"), optional leading whitespace, the field value, and
   optional trailing whitespace.

### PR DESCRIPTION
**Pull request by kyoshidajp**
Fixes #1510.

### CODE REVIEW

File: src/urllib3/util/retry.py [hunk 1]
```diff
@@ -179,7 +179,7 @@ def __init__(self, total=10, connect=None, read=None, redirect=None, status=None
         self.raise_on_status = raise_on_status
         self.history = history or tuple()
         self.respect_retry_after_header = respect_retry_after_header
-        self.remove_headers_on_redirect = remove_headers_on_redirect
+        self.remove_headers_on_redirect = [h.lower() for h in remove_headers_on_redirect]
```
Comment (sethmlarson) [inline src/urllib3/util/retry.py]: I think we should change this structure into a frozenset to protect users from adding headers not through the constructor if we're going to enforce lowercase-only.

File: src/urllib3/poolmanager.py [hunk 2]
```diff
@@ -342,8 +342,12 @@ def urlopen(self, method, url, redirect=True, **kw):
         # conn.is_same_host() which may use socket.gethostbyname() in the future.
         if (retries.remove_headers_on_redirect
                 and not conn.is_same_host(redirect_location)):
-            for header in retries.remove_headers_on_redirect:
```
Comment (sethmlarson) [inline src/urllib3/poolmanager.py]: Let's keep some of the original structure of this logic by instead looping through all the header names with `six.viewkeys()` (so we don't have to make a copy) and then seeing if that header name is within the `remove_headers_on_redirect` frozenset.

File: test/with_dummyserver/test_poolmanager.py [hunk 3]
```diff
@@ -123,6 +123,17 @@ def test_redirect_cross_host_remove_headers(self):
...
+        r = http.request('GET', '%s/redirect' % self.base_url,
+                         fields={'target': '%s/headers' % self.base_url_alt},
+                         headers={'authorization': 'foo'})
+
+        self.assertEqual(r.status, 200)
+
+        data = json.loads(r.data.decode('utf-8'))
+
+        self.assertNotIn('authorization', data)
+        self.assertNotIn('Authorization', data)
```
Comment (sethmlarson) [inline test/with_dummyserver/test_poolmanager.py]: Let's also add a new test for an additional header (eg `X-API-Secret`) and seeing that that header is removed.

**Comment by sethmlarson** [comment]
Also don't worry about force-pushing, we squash all commits. :)

File: test/with_dummyserver/test_poolmanager.py (hunk 3 above)
Comment (sethmlarson) [reply test/with_dummyserver/test_poolmanager.py]: `test_redirect_cross_host_set_removed_headers`

File: src/urllib3/util/retry.py (hunk 1 above)
Comment (kyoshidajp) [reply src/urllib3/util/retry.py]: You mean we should remove `remove_headers_on_redirect` parameter from the constructor?

File: src/urllib3/poolmanager.py (hunk 2 above)
Comment (kyoshidajp) [reply src/urllib3/poolmanager.py]: I changed by 4357014. But made a copy by `.copy()` due to getting `RuntimeError: dictionary changed size during iteration` if without `.copy()`.

File: src/urllib3/util/retry.py (hunk 1 above)
Comment (sethmlarson) [reply src/urllib3/util/retry.py]: No, I mean converting the iterable we receive from remove_headers_on_redirect into a frozenset. Something like `self.remove_headers_on_redirect = frozenset([h.lower() for h in remove_headers_on_redirect])`

**Comment by kyoshidajp** [comment]
Thanks a lot!

**Comment by kyoshidajp** [comment]
@sethmlarson Could you review again?

**Comment by sethmlarson** [comment]
Thanks for this!

**Pull request merged**